# Проект по SQL - Анализ базы данных для чтения книг по подписке

Компания решила купить крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель проекта** - проанализировать базу данных. В базе данных информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

**Описание данных**

_Таблица books_

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

_Таблица authors_

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

_Таблица publishers_

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

_Таблица ratings_

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

_Таблица reviews_

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# полчучаем данные таблицы books
query_books = '''
            SELECT * 
            FROM books
        '''

In [5]:
# полчучаем данные таблицы authors
query_authors = '''
            SELECT * 
            FROM authors
        '''

In [6]:
# полчучаем данные таблицы publishers
query_publishers = '''
            SELECT * 
            FROM publishers
        '''

In [7]:
# полчучаем данные таблицы ratings
query_ratings = '''
            SELECT * 
            FROM ratings
        '''

In [8]:
# полчучаем данные таблицы reviews
query_reviews = '''
            SELECT * 
            FROM reviews
        '''

In [9]:
# подключения к базам данных для Sqlite
books = pd.io.sql.read_sql(query_books, con = engine)
authors = pd.io.sql.read_sql(query_authors, con = engine)
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
reviews = pd.io.sql.read_sql(query_reviews, con = engine)

In [10]:
# посмотрим общую информацию о данных
data = [books, authors, publishers, ratings, reviews]

for i in data:
    display(i.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

In [11]:
# посмотрим на 5 первых значений
for i in data:
    display(i.head())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Вывод**

В новой базе будет содержатся информация о 1000 книгах, 636 авторов, 340 издательств и 6456 пользовательских обзорах.

**Задание 1**

1. Посчитаем количество книг, которые вышли после 1 января 2000 года

In [12]:
request  = '''
            SELECT COUNT(book_id) 
            FROM books
            WHERE publication_date > '2000-01-01'
       '''
pd.read_sql(request , con = engine)

count
0    819

In [13]:
request  = '''
            SELECT COUNT(book_id) 
            FROM books
            WHERE publication_date >= '2000-01-01'
       '''
pd.read_sql(request, con = engine)

count
0    821

**Вывод**

После 1 января 2000 года вышло 821 книг. Хотела пояснить, нелогично икслючать 1 января 2020г т.к. скорее нужна была информация за год, который закончился или нет.

**Задание 2**

2. Посчитаем, для каждой книги количество обзоров и среднюю оценку

In [21]:
request  = '''
         SELECT books.book_id, 
                books.title, 
                COUNT(DISTINCT reviews.review_id) AS count_reviews, -- выводим кол-во обзоров и переименовываем count_reviews 
                AVG(ratings.rating) AS avg_ratings -- выводим средний рейтинг и переименовываем avg_ratings        
         FROM books
         LEFT JOIN ratings ON books.book_id = ratings.book_id -- присоединяем к таб. books таб. ratings по ключу book_id
         LEFT JOIN reviews ON books.book_id = reviews.book_id -- присоединяем к таб. books таб. reviews по ключу book_id
         GROUP BY books.book_id   -- группируем по полю book_id
         ORDER BY count_reviews DESC -- сортируем по убыванию кол-во обзоров 
    '''
pd.read_sql(request, con = engine)

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     count_reviews  avg_ratings  
0                7     3.662500  
1                6     3.977273  
2                6     4.206897  
3                6     4.414634  
4                6     4.081081  
..             ...          ...  
995              0     3.666667  
996              0     3.000000  
997              0     5.000000  
998              0     4.000000  
999              0     4.000000  

[1000 rows x 4 columns]

**Вывод**

Самое большое количество обзоров у книги 'Twilight (Twilight #1)'  - 7 шт, средний рейтинг ~ 3,7.  Есть книги которые не имеют обзоров.

**Задание 3**

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц (так вы исключим из анализа брошюры)

In [18]:
request  = '''
         SELECT publishers.publisher, 
                COUNT(books.book_id) AS count_book -- выводим кол-во книг и переименовываем count_book 
         FROM publishers
         LEFT JOIN books ON publishers.publisher_id = books.publisher_id -- присоединяем к таб.publisher таб.books по ключу publisher_id
         WHERE num_pages > 50 -- запишем условие толщина книг больше 50 страниц 
         GROUP BY publishers.publisher
         ORDER BY count_book DESC -- сортируем по убыванию кол-ву книг
         LIMIT 1
            
    '''
pd.read_sql(request, con = engine)

publisher  count_book
0  Penguin Books          42

**Вывод**

Издательство 'Penguin Books'  выпустило наибольшее число книг - 42 шт толщиной более 50 страниц.

**Задание 4**

Определим автора с самой высокой средней оценкой книг — учитывая только книги с 50 и более оценками.

In [19]:
request = '''
        SELECT 
             authors.author, 
             AVG(ratings.rating) AS avg_rating -- выводим средний рейтинг книг и переименовываем avg_rating
        FROM authors
        LEFT JOIN books ON authors.author_id = books.author_id -- присоединяем к таб.authors таб.authors по ключу author_id
        LEFT JOIN ratings ON books.book_id = ratings.book_id -- присоединяем к таб.books таб.ratings по ключу book_id 
        WHERE books.book_id IN (SELECT books.book_id
                                FROM books    
                                LEFT JOIN ratings ON books.book_id  = ratings.book_id     
                                GROUP BY books.book_id
                                HAVING COUNT(ratings.rating_id) >= 50)
        GROUP BY authors.author  
        ORDER BY avg_rating DESC
        LIMIT 1  
            
    '''
pd.read_sql(request, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

**Вывод** 

Автор с самой высокой средней оценкой книг при условии использования книги с 50 и более оценками - Mary GrandPré. 

**Задание 5**

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [20]:
request = '''
WITH
-- первый подзапрос с псевдонимом ratings_count
   ratings_count AS (SELECT ratings.username,
                            COUNT(rating_id)  -- выводим кол-во оценок
                     FROM ratings  
                     GROUP BY ratings.username 
                     HAVING COUNT(rating_id) > 50),
-- второй подзапрос с псевдонимом reviews_count                   
   reviews_count AS (SELECT reviews.username,
                            COUNT(review_id) AS count_reviews -- выводим кол-во обзоров
                     FROM reviews 
                     GROUP BY reviews.username)
-- основной запрос, в котором указываем псевдонимы для подзапросов
SELECT AVG(count_reviews) -- ср.кол-во обзоров
FROM ratings_count
LEFT JOIN reviews_count ON ratings_count.username = reviews_count.username -- присоединяем псевдоним ratings_count к ratings_count по ключу username  

'''
pd.read_sql(request, con = engine).round()

avg
0  24.0

**Вывод**

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

**Выводы исследования**:

1. После 1 января 2000 года вышло 821 книг.
2. Самое большое количество обзоров у книги 'Twilight (Twilight #1)' - 1120 шт.
3. Издательство 'Penguin Books'  выпустило наибольшее число книг - 42 шт толщиной более 50 страниц.
4. Автор с самой высокой средней оценкой книг при условии использования книги с 50 и более оценками - Mary GrandPré.
5. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.